### PR TITLE
Azure Functions - Publishing enhancements

### LINKED ISSUE #585: Azure Functions: choose OS (Windows/Linux) when creating new function app
Add the ability to choose OS (Windows/Linux) when creating new function app

### LINKED ISSUE #586: Azure Functions: Set correct worker runtime when publishing
Azure Functions: Set correct worker runtime when publishing. Now always sets `dotnet`, even for isolated worker instances.

### PR DESCRIPTION
**Pull request by maartenba**
Fixes #585
Fixes #586

Todo:
* [x] Check using existing hosting plan
* [x] Make sure pricing plans are not listed multiple times
* [x] Support consumption plan for Linux? - https://twitter.com/maartenballiauw/status/1514975559746523137 and https://github.com/projectkudu/kudu/issues/3362

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/FunctionAppRunState.kt [hunk 1]
```diff
@@ -82,6 +82,30 @@ class FunctionAppRunState(project: Project, private val myModel: FunctionAppSett
                 AzureRiderSettings.VALUE_COLLECT_ARTIFACTS_TIMEOUT_MINUTES_DEFAULT) * 60000L
 
         val app = getOrCreateFunctionAppFromConfiguration(myModel.functionAppModel, processHandler)
+
+        if (myModel.functionAppModel.isCreatingNewApp && app is FunctionApp) {
```
Comment (maartenba) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/FunctionAppRunState.kt:86]: This section updates the runtime settings in Azure (only when creating it from Rider, as we do with web apps).

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/ui/FunctionAppPublishComponent.kt [hunk 2]
```diff
@@ -168,6 +180,27 @@ class FunctionAppPublishComponent(private val lifetime: Lifetime,
...
+
+        // Set runtime stack based on project config
+        val publishableProject = model.publishableProject
+        if (publishableProject != null && publishableProject.isDotNetCore) {
+            val functionLocalSettings = FunctionLocalSettingsUtil.readFunctionLocalSettings(project, File(publishableProject.projectFilePath).parent)
+            val workerRuntime = functionLocalSettings?.values?.workerRuntime ?: FunctionsWorkerRuntime.DotNetDefault
+
+            val coreToolsVersion = FunctionsCoreToolsMsBuild.requestAzureFunctionsVersion(project, publishableProject.projectFilePath) ?: "V4"
+            val netCoreVersion = getProjectNetCoreFrameworkVersion(publishableProject)
+            model.functionRuntimeStack = FunctionRuntimeStack(
```
Comment (maartenba) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/ui/FunctionAppPublishComponent.kt:201]: Here we build `FunctionRuntimeStack`, based on the project.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/FunctionAppRunState.kt [hunk 3]
```diff
@@ -125,6 +126,30 @@ class FunctionAppRunState(project: Project, private val myModel: FunctionAppSett
         return FunctionAppDeployResult(app, database)
     }
 
+    private fun tryConfigureAzureFunctionRuntimeStack(app: WebAppBase, subscriptionId: String, processHandler: RunProcessHandler) {
+        if (app !is FunctionApp) return
+
+        val functionRuntimeStack = myModel.functionAppModel.functionRuntimeStack
```
Comment (maartenba) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/FunctionAppRunState.kt:132]: This method updates the runtime settings in Azure.

**Comment by maartenba** [comment]
@sdubov Ready to go :)

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/FunctionAppRunState.kt (hunk 3 above)
Comment (sdubov) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/FunctionAppRunState.kt:132]: This logic raise one more request to the server. Do we have an option to configure an app on create request using the same runtime configuration logic? Or is this API available after app creation only?

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/ui/FunctionAppPublishComponent.kt [hunk 4]
```diff
@@ -168,6 +180,31 @@ class FunctionAppPublishComponent(private val lifetime: Lifetime,
...
         }
 
+        if (pnlFunctionAppSelector.isCreateNew) {
+            if (pnlCreateFunctionApp.pnlOperatingSystem.isWindows) {
+                model.operatingSystem = OperatingSystem.WINDOWS
+            } else {
+                model.operatingSystem = OperatingSystem.LINUX
+            }
+        } else {
+            model.operatingSystem = selectedApp?.operatingSystem() ?: OperatingSystem.WINDOWS
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/ui/FunctionAppPublishComponent.kt:190]: Do we use Windows by default? or throw an exception here?

**Review by sdubov — APPROVED** [review]
LGTM. One question about optimisation of requests for deploy.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/FunctionAppRunState.kt (hunk 3 above)
Comment (maartenba) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/FunctionAppRunState.kt:132]: Correct. If you update your project TFM or change from isolated to default worker, we have to make sure it all works.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/ui/FunctionAppPublishComponent.kt (hunk 4 above)
Comment (maartenba) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/runner/functionapp/config/ui/FunctionAppPublishComponent.kt:190]: Windows by default, like we do in web apps.

**Pull request merged**